### PR TITLE
Make target type preselection less aggressive

### PR DESCRIPTION
**Pull request by rchande**
Fixes https://github.com/dotnet/roslyn/issues/14085

### CODE REVIEW

**Comment by rchande** [comment]
@dotnet/roslyn-ide pls review

**Comment by CyrusNajmabadi** [comment]
Your title doesn't seem to address the bug reported here: Type preselection takes priority over exact case matches in C# 

Basically, exact case match should always succeed, regardless of what the MRU says.   Type preselection should only be used to weight results in a specific goodness category.  i.e. if you have an exact match that should still beat a type-preselected prefix match.

Is that what this PR accomplishes?  Or is this just an MRU change?

**Comment by rchande** [comment]
@CyrusNajmabadi (Changed title) This fixes
https://github.com/dotnet/roslyn/issues/14056 and
https://github.com/dotnet/roslyn/issues/14085

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs [hunk 1]
```diff
@@ -299,6 +304,23 @@ private Boolean IsAfterDot(Model model, ITextSnapshot textSnapshot, Dictionary<T
                     }
                 }
 
+                if (GetRecentItemIndex(recentItems, bestItem) < 1)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: Probably much clearer as <= 0.  Or, just "if (recentItems.IndexOf(bestItem.DisplayText) >= 0)"

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs [hunk 2]
```diff
@@ -299,6 +304,23 @@ private Boolean IsAfterDot(Model model, ITextSnapshot textSnapshot, Dictionary<T
...
+                {
+                    return bestItem;
+                }
+
+                for (int i = 1, n = chosenItems.Length; i < n; i++)
+                {
+                    var chosenItem = chosenItems[i];
+                    var mruIndex1 = bestItem.Rules.MatchPriority;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: i don't understnad these variable names.  There's also serious lack of comments explaining what scenario this tries to ensure.

File: src/Features/Core/Portable/Completion/CompletionHelper.cs [hunk 3]
```diff
@@ -184,10 +184,15 @@ public int CompareItems(CompletionItem item1, CompletionItem item2, string filte
...
-            var diff = item1.Rules.MatchPriority - item2.Rules.MatchPriority;
-            if (diff != 0)
+            if (item1.Rules.MatchPriority == MatchPriority.Preselect && 
+                item2.Rules.MatchPriority != MatchPriority.Preselect)
             {
-                return -diff;
+                return -1;
+            }
+            else if (item1.Rules.MatchPriority != MatchPriority.Preselect &&
+                item2.Rules.MatchPriority == MatchPriority.Preselect)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: align appropriately with above.

File: src/Features/Core/Portable/Completion/CompletionHelper.cs [hunk 4]
```diff
@@ -229,15 +234,17 @@ private int CompareMatches(PatternMatch match1, PatternMatch match2, CompletionI
...
             }
 
-            // Now, after comparing matches, check if an item wants to be preselected.  If so,
-            // we prefer that.  i.e. say the user has typed 'f' and we have the items 'foo' 
-            // and 'False' (with the latter being 'Preselected').  Both will be a prefix match.
-            // And because we are ignoring case, neither will be seen as better.  Now, because
-            // 'False' is preselected we pick it even though 'foo' matches 'f' case sensitively.
-            diff = item2.Rules.MatchPriority - item1.Rules.MatchPriority;
-            if (diff != 0)
+            // If they both seemed just as good, but they differ on preselection, then
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: this logic seems duplicated.  Extract out into helper.  Also, why do we have it in two places?  Once it was checked once, why is there another codepath that checks it elsewhere?

**Comment by jasonmalinowski** [comment]
Should this be in RC? This bug is _really_ annoying.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs (hunk 2 above)
Comment (rchande) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: Fixed

File: src/Features/Core/Portable/Completion/CompletionHelper.cs (hunk 3 above)
Comment (rchande) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: Fixed.

File: src/Features/Core/Portable/Completion/CompletionHelper.cs (hunk 4 above)
Comment (rchande) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: Removed the redundant check from CompareItems--it calls CompareMatches which already does this check.

**Comment by rchande** [comment]
Ping @dotnet/roslyn-ide -- Rebased against dev15-rc branch

**Comment by rchande** [comment]
Tag @MattGertz for Ask Mode approval. This fixes some highly irritating regressions that were introduced by "target type preselection".

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs [hunk 5]
```diff
@@ -444,7 +471,7 @@ public FilterResult(PresentationItem presentationItem, string filterText, bool m
                         return true;
                     }
 
-                    if (!recentItems.IsDefault && GetRecentItemIndex(recentItems, item) < 0)
+                    if (!recentItems.IsDefault && GetRecentItemIndex(recentItems, item) < 1)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller.Session_FilterModel.cs]: <= 0

File: src/Features/Core/Portable/Completion/CompletionHelper.cs [hunk 6]
```diff
@@ -229,15 +221,17 @@ private int CompareMatches(PatternMatch match1, PatternMatch match2, CompletionI
...
+            // item1 is better if it is preselected, otherwise it is worse.
+            if (item1.Rules.MatchPriority == MatchPriority.Preselect &&
+                item2.Rules.MatchPriority != MatchPriority.Preselect)
             {
-                return diff;
+                return -1;
+            }
+            else if (item1.Rules.MatchPriority != MatchPriority.Preselect &&
+                item2.Rules.MatchPriority == MatchPriority.Preselect)
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: indentation.

**Pull request merged**